src/tags/models.py

```diff
@@ -34,21 +34,66 @@ def get_tags_from_user(user):
     return (quest_tags | badge_tags).distinct()
 
 
+def get_quest_by_tag(user, tags):
+    """
+    Returns queryset of quest objects that are related to tag and user
+
+    Args:
+        user (UserModel): user object
+        tags (list[str], Queryset<Tag>): a queryset or list of Tag objects
+
+    Returns:
+        QS: queryset of quest objects that are related to tag and user
+    """
+    # get model through here to prevent circular imports
+    Quest = apps.get_model("quest_manager", "Quest")
+    QuestSubmission = apps.get_model("quest_manager", "QuestSubmission")
+
+    id_list = QuestSubmission.objects.all_approved(user, active_semester_only=True).filter(
+        is_completed=True, 
+        quest__tags__name__in=list(tags)
+    ).exclude(do_not_grant_xp=True).distinct().values_list('quest__id', flat=True)
+
+    return Quest.objects.filter(id__in=id_list)
+
+
+def get_badge_by_tag(user, tags):
```

Comment:
Problem with not having a test for this, is that its not the same code.  This code could be change and the other test wont fail.  However there is a solution to prevent duplicate tests.  Put the common code in its own method and then have both these methods call the common one, and the common method used by both can be fully tested.

If you can't put it in a common method, well, then they aren't the same and need their own tests.

Same with the other method.

Reply:
Taking a closer look, seems like you should put the common parts into new common methods that are tested (to ensure the proper values are returned)

Then each individual method here can have their own much simpler tests.  They can assume the proper values are returned, but might need to test that the aggregate was added in one, and that the values list was added in the other.

Reply:
added some tests